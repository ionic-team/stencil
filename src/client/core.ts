import { AppGlobal } from '../util/interfaces';
import { Build } from '../util/build-conditionals';
import { createPlatformClient } from './platform-client';


const App: AppGlobal = (<any>window)[appNamespace] = (<any>window)[appNamespace] || {};

const plt = createPlatformClient(Context, App, window, document, publicPath, hydratedCssClass);


if (Build.es5) {
  plt.registerComponents(App.components).forEach(cmpMeta => {
    // es5 way of extending HTMLElement
    function HostElement(self: any) {
      return HTMLElement.call(this, self);
    }

    HostElement.prototype = Object.create(
      HTMLElement.prototype,
      { constructor: { value: HostElement, configurable: true } }
    );

    plt.defineComponent(cmpMeta, HostElement);
  });

} else {
  // es6 class extends HTMLElement
  plt.registerComponents(App.components).forEach(cmpMeta =>
    plt.defineComponent(cmpMeta, class extends HTMLElement {}));
}
