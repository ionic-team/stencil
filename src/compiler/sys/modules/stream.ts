export const Stream = function () {
  /**/
};

export default {
  Stream,
};
