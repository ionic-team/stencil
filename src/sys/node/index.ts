export { createNodeLogger } from './node-logger';
export { setupNodeProcess } from './node-setup-process';
export { createNodeSys } from './node-sys';
