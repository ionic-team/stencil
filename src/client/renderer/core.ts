/**
 * Virtual DOM patching algorithm based on Snabbdom by
 * Simon Friis Vindum (@paldepind)
 * Licensed under the MIT License
 * https://github.com/snabbdom/snabbdom/blob/master/LICENSE
 *
 * Modified for Ionic's Web Component Renderer
 * /


/* global module, document, Node */
import { isArray, isDef, isUndef, isString, isStringOrNumber } from '../../util/helpers';
import { HostContentNodes, PlatformApi, ProxyElement, RendererApi, VNode, VNodeData, Key } from '../../util/interfaces';
import { vnode } from './vnode';

import { updateAttrs } from './modules/attributes';
import { updateClass } from './modules/class';
import { updateStyle } from './modules/style';
import { updateEventListeners } from './modules/eventlisteners';
import { updateProps } from './modules/props';
export { VNode, VNodeData, vnode };
export { h } from './h';

type VNodeQueue = Array<VNode>;

const emptyNode = vnode('', {}, [], undefined, undefined);

function sameVnode(vnode1: VNode, vnode2: VNode): boolean {
  return vnode1.vkey === vnode2.vkey && vnode1.sel === vnode2.sel;
}

function isVnode(vnode: any): vnode is VNode {
  return vnode.sel !== undefined || vnode.elm !== undefined;
}

type KeyToIndexMap = {[key: string]: number};

function createKeyToOldIdx(children: Array<VNode>, beginIdx: number, endIdx: number): KeyToIndexMap {
  let i: number, map: KeyToIndexMap = {}, key: Key, ch;
  for (i = beginIdx; i <= endIdx; ++i) {
    ch = children[i];
    if (ch != null) {
      key = (<VNode>ch).vkey;
      if (key !== undefined) map[key] = i;
    }
  }
  return map;
}


export function Renderer(api: PlatformApi): RendererApi {

  function emptyNodeAt(elm: Element) {
    const id = elm.id ? '#' + elm.id : '';
    const c = elm.className ? '.' + elm.className.split(' ').join('.') : '';
    return vnode(api.$tagName(elm) + id + c, {}, [], undefined, elm);
  }

  function createRmCb(childElm: Node) {
    return function rmCb() {
      const parent = api.$parentNode(childElm);
      api.$removeChild(parent, childElm);
    };
  }

  function createElm(vnode: VNode, insertedVnodeQueue: VNodeQueue, parentElm: Node, hostContentNodes: HostContentNodes): Node {
    let i: any, data = vnode.vdata;
    if (data !== undefined) {
      if (isDef(i = data.hook) && isDef(i = i.init)) {
        i(vnode);
        data = vnode.vdata;
      }
    }
    let children = vnode.vchildren, sel = vnode.sel;
    if (sel === '!') {
      if (isUndef(vnode.vtext)) {
        vnode.vtext = '';
      }
      vnode.elm = api.$createComment(vnode.vtext as string);

    } else if (hostContentNodes && sel === 'slot') {
      // special case for manually relocating host content nodes
      // there their new home in either a named slot or the default slot
      let namedSlot = (data.attrs && data.attrs.name);
      let slotNodes: Node[];

      if (namedSlot) {
        slotNodes = hostContentNodes.$namedSlots && hostContentNodes.$namedSlots[namedSlot];

      } else {
        slotNodes = hostContentNodes.$defaultSlot;
      }

      if (slotNodes) {
        for (let nodeIndex = 0; nodeIndex < slotNodes.length; nodeIndex++) {
          // remove the host content node from it's original parent node
          (<ProxyElement>slotNodes[nodeIndex]).$tmpDisconnected = true;
          api.$removeChild(slotNodes[nodeIndex].parentNode, slotNodes[nodeIndex]);

          if (nodeIndex === slotNodes.length - 1) {
            // return the last node that gets appended
            // like any other Node that was created
            return slotNodes[nodeIndex];
          }

          // relocate the node to it's new home
          api.$appendChild(parentElm, slotNodes[nodeIndex]);
          delete (<ProxyElement>slotNodes[nodeIndex]).$tmpDisconnect;
        }
      }

    } else if (sel !== undefined) {
      // Parse selector
      const hashIdx = sel.indexOf('#');
      const dotIdx = sel.indexOf('.', hashIdx);
      const hash = hashIdx > 0 ? hashIdx : sel.length;
      const dot = dotIdx > 0 ? dotIdx : sel.length;
      const tag = hashIdx !== -1 || dotIdx !== -1 ? sel.slice(0, Math.min(hash, dot)) : sel;
      const elm = vnode.elm = isDef(data) && isDef(i = (data as VNodeData).ns) ? api.$createElementNS(i, tag)
                                                                               : api.$createElement(<any>tag);
      if (hash < dot) elm.setAttribute('id', sel.slice(hash + 1, dot));
      if (dotIdx > 0) elm.setAttribute('class', sel.slice(dot + 1).replace(/\./g, ' '));

      updateAttrs(emptyNode, vnode);
      updateClass(emptyNode, vnode);
      updateStyle(emptyNode, vnode);
      updateEventListeners(emptyNode, vnode);
      updateProps(emptyNode, vnode);

      data.ref && data.ref(elm);

      if (isArray(children)) {
        for (i = 0; i < children.length; ++i) {
          let ch = children[i];
          if (ch != null) {
            ch = <any>createElm(ch as VNode, insertedVnodeQueue, elm, hostContentNodes);
            if (ch) {
              api.$appendChild(elm, <any>ch);
            }
          }
        }
      } else if (isStringOrNumber(vnode.vtext)) {
        api.$appendChild(elm, api.$createTextNode(vnode.vtext));
      }
      i = (vnode.vdata as VNodeData).hook; // Reuse variable
      if (isDef(i)) {
        if (i.create) i.create(emptyNode, vnode);
        if (i.insert) insertedVnodeQueue.push(vnode);
      }
    } else {
      vnode.elm = api.$createTextNode(vnode.vtext as string);
    }
    return vnode.elm;
  }

  function addVnodes(parentElm: Node,
                     before: Node | null,
                     vnodes: Array<VNode>,
                     startIdx: number,
                     endIdx: number,
                     insertedVnodeQueue: VNodeQueue,
                     hostContentNodes: HostContentNodes) {
    for (; startIdx <= endIdx; ++startIdx) {
      const ch = vnodes[startIdx];
      if (ch != null) {
        api.$insertBefore(parentElm, createElm(ch, insertedVnodeQueue, parentElm, hostContentNodes), before);
      }
    }
  }

  function removeVnodes(parentElm: Node,
                        vnodes: Array<VNode>,
                        startIdx: number,
                        endIdx: number): void {
    for (; startIdx <= endIdx; ++startIdx) {
      let i: any, rm: () => void, ch = vnodes[startIdx];
      if (ch != null) {
        if (isDef(ch.sel)) {
          invokeDestroyHook(ch);
          rm = createRmCb(ch.elm as Node);
          if (isDef(i = ch.vdata) && isDef(i = i.hook) && isDef(i = i.remove)) {
            i(ch, rm);
          } else {
            rm();
          }
        } else { // Text node
          api.$removeChild(parentElm, ch.elm as Node);
        }
      }
    }
  }

  function updateChildren(parentElm: Node,
                          oldCh: Array<VNode>,
                          newCh: Array<VNode>,
                          insertedVnodeQueue: VNodeQueue,
                          hostContentNodes: HostContentNodes) {
    let oldStartIdx = 0, newStartIdx = 0;
    let oldEndIdx = oldCh.length - 1;
    let oldStartVnode = oldCh[0];
    let oldEndVnode = oldCh[oldEndIdx];
    let newEndIdx = newCh.length - 1;
    let newStartVnode = newCh[0];
    let newEndVnode = newCh[newEndIdx];
    let oldKeyToIdx: any;
    let idxInOld: number;
    let elmToMove: VNode;
    let before: any;

    while (oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx) {
      if (oldStartVnode == null) {
        oldStartVnode = oldCh[++oldStartIdx]; // Vnode might have been moved left
      } else if (oldEndVnode == null) {
        oldEndVnode = oldCh[--oldEndIdx];
      } else if (newStartVnode == null) {
        newStartVnode = newCh[++newStartIdx];
      } else if (newEndVnode == null) {
        newEndVnode = newCh[--newEndIdx];
      } else if (sameVnode(oldStartVnode, newStartVnode)) {
        patchVnode(oldStartVnode, newStartVnode, insertedVnodeQueue, hostContentNodes);
        oldStartVnode = oldCh[++oldStartIdx];
        newStartVnode = newCh[++newStartIdx];
      } else if (sameVnode(oldEndVnode, newEndVnode)) {
        patchVnode(oldEndVnode, newEndVnode, insertedVnodeQueue, hostContentNodes);
        oldEndVnode = oldCh[--oldEndIdx];
        newEndVnode = newCh[--newEndIdx];
      } else if (sameVnode(oldStartVnode, newEndVnode)) { // Vnode moved right
        patchVnode(oldStartVnode, newEndVnode, insertedVnodeQueue, hostContentNodes);
        api.$insertBefore(parentElm, oldStartVnode.elm as Node, api.$nextSibling(oldEndVnode.elm as Node));
        oldStartVnode = oldCh[++oldStartIdx];
        newEndVnode = newCh[--newEndIdx];
      } else if (sameVnode(oldEndVnode, newStartVnode)) { // Vnode moved left
        patchVnode(oldEndVnode, newStartVnode, insertedVnodeQueue, hostContentNodes);
        api.$insertBefore(parentElm, oldEndVnode.elm as Node, oldStartVnode.elm as Node);
        oldEndVnode = oldCh[--oldEndIdx];
        newStartVnode = newCh[++newStartIdx];
      } else {
        if (oldKeyToIdx === undefined) {
          oldKeyToIdx = createKeyToOldIdx(oldCh, oldStartIdx, oldEndIdx);
        }
        idxInOld = oldKeyToIdx[newStartVnode.vkey as string];
        if (isUndef(idxInOld)) { // New element
          api.$insertBefore(parentElm, createElm(newStartVnode, insertedVnodeQueue, parentElm, hostContentNodes), oldStartVnode.elm as Node);
          newStartVnode = newCh[++newStartIdx];
        } else {
          elmToMove = oldCh[idxInOld];
          if (elmToMove.sel !== newStartVnode.sel) {
            api.$insertBefore(parentElm, createElm(newStartVnode, insertedVnodeQueue, parentElm, hostContentNodes), oldStartVnode.elm as Node);
          } else {
            patchVnode(elmToMove, newStartVnode, insertedVnodeQueue, hostContentNodes);
            oldCh[idxInOld] = undefined as any;
            api.$insertBefore(parentElm, (elmToMove.elm as Node), oldStartVnode.elm as Node);
          }
          newStartVnode = newCh[++newStartIdx];
        }
      }
    }
    if (oldStartIdx > oldEndIdx) {
      before = newCh[newEndIdx + 1] == null ? null : newCh[newEndIdx + 1].elm;
      addVnodes(parentElm, before, newCh, newStartIdx, newEndIdx, insertedVnodeQueue, hostContentNodes);
    } else if (newStartIdx > newEndIdx) {
      removeVnodes(parentElm, oldCh, oldStartIdx, oldEndIdx);
    }
  }

  function patchVnode(oldVnode: VNode, vnode: VNode, insertedVnodeQueue: VNodeQueue, hostContentNodes: HostContentNodes) {
    if (!hostContentNodes && oldVnode.sel === 'slot') {
      return;
    }

    let i: any, hook: any;
    if (isDef(i = vnode.vdata) && isDef(hook = i.hook) && isDef(i = hook.prepatch)) {
      i(oldVnode, vnode);
    }

    const elm = vnode.elm = (oldVnode.elm as Node);
    let oldCh = oldVnode.vchildren;
    let ch = vnode.vchildren;
    if (oldVnode === vnode) return;
    if (vnode.vdata !== undefined) {
      updateAttrs(oldVnode, vnode);
      updateClass(oldVnode, vnode);
      updateStyle(emptyNode, vnode);
      updateEventListeners(oldVnode, vnode);
      updateProps(oldVnode, vnode);

      i = vnode.vdata.hook;
      if (isDef(i) && isDef(i = i.update)) i(oldVnode, vnode);
    }
    if (isUndef(vnode.vtext)) {
      if (isDef(oldCh) && isDef(ch)) {
        if (oldCh !== ch) {
          updateChildren((<HTMLElement>elm).shadowRoot || elm, oldCh as Array<VNode>, ch as Array<VNode>, insertedVnodeQueue, hostContentNodes);
        }

      } else if (isDef(ch)) {
        if (isDef(oldVnode.vtext)) api.$setTextContent(elm, '');
        addVnodes(elm, null, ch as Array<VNode>, 0, (ch as Array<VNode>).length - 1, insertedVnodeQueue, null);
      } else if (isDef(oldCh)) {
        removeVnodes(elm, oldCh as Array<VNode>, 0, (oldCh as Array<VNode>).length - 1);
      } else if (isDef(oldVnode.vtext)) {
        api.$setTextContent(elm, '');
      }
    } else if (oldVnode.vtext !== vnode.vtext) {
      api.$setTextContent(elm, vnode.vtext as string);
    }
    if (isDef(hook) && isDef(i = hook.postpatch)) {
      i(oldVnode, vnode);
    }
  }

  return function patch(oldVnode: VNode | Element, vnode: VNode, hostContentNodes?: HostContentNodes): VNode {
    let elm: Node, parent: Node;
    const insertedVnodeQueue: VNodeQueue = [];

    if (!isVnode(oldVnode)) {
      oldVnode = emptyNodeAt(oldVnode);
    }

    if (vnode.elm || sameVnode(oldVnode, vnode)) {
      patchVnode(oldVnode, vnode, insertedVnodeQueue, hostContentNodes);

    } else {
      elm = oldVnode.elm as Node;
      parent = api.$parentNode(elm);

      createElm(vnode, insertedVnodeQueue, parent, hostContentNodes);

      if (parent !== null) {
        api.$insertBefore(parent, vnode.elm as Node, api.$nextSibling(elm));
        removeVnodes(parent, [oldVnode], 0, 0);
      }
    }
    return vnode;
  };
}


export function invokeDestroyHook(vnode: VNode) {
  let i: any, j: number, data = vnode.vdata;
  if (data !== undefined) {
    updateEventListeners(vnode);
    if (vnode.vchildren !== undefined) {
      for (j = 0; j < vnode.vchildren.length; ++j) {
        i = vnode.vchildren[j];
        if (i != null && !isString(i)) {
          invokeDestroyHook(i);
        }
      }
    }
  }
}
