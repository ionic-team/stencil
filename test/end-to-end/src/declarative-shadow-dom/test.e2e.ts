import { Readable } from 'node:stream';

import { newE2EPage } from '@stencil/core/testing';

import { CarData } from '../car-list/car-data';

const vento = new CarData('VW', 'Vento', 2024);
const beetle = new CarData('VW', 'Beetle', 2023);

async function readableToString(readable: Readable) {
  return new Promise((resolve, reject) => {
    let data = '';

    readable.on('data', (chunk) => {
      data += chunk;
    });

    readable.on('end', () => {
      resolve(data);
    });

    readable.on('error', (err) => {
      reject(err);
    });
  });
}

// @ts-ignore may not be existing when project hasn't been built
type HydrateModule = typeof import('../../hydrate');
let renderToString: HydrateModule['renderToString'];
let streamToString: HydrateModule['streamToString'];
let hydrateDocument: HydrateModule['hydrateDocument'];

describe('renderToString', () => {
  beforeAll(async () => {
    // @ts-ignore may not be existing when project hasn't been built
    const mod = await import('../../hydrate');
    renderToString = mod.renderToString;
    streamToString = mod.streamToString;
    hydrateDocument = mod.hydrateDocument;
  });

  it('resolves to a Promise<HydrateResults> by default', async () => {
    const renderedString = renderToString('<div>Hello World</div>');
    expect(typeof renderedString.then).toBe('function');
    // this is a type assertion to verify that the promise resolves to a HydrateResults object
    renderedString.then((result) => result.html);

    const renderedDocument = hydrateDocument('<div>Hello World</div>');
    expect(typeof renderedDocument.then).toBe('function');
    // this is a type assertion to verify that the promise resolves to a HydrateResults object
    renderedDocument.then((result) => result.html);
  });

  it('can render a simple dom node', async () => {
    const { html } = await renderToString('<div>Hello World</div>');
    expect(html).toContain('<body><div>Hello World</div></body>');
  });

  it('can render a simple dom node (sync)', async () => {
    const input = '<div>Hello World</div>';
    const renderedHTML = '<body><div>Hello World</div></body>';
    const stream = renderToString(input, {}, true);
    expect(await readableToString(stream)).toContain(renderedHTML);
    expect(await readableToString(streamToString(input))).toContain(renderedHTML);
  });

  it('renders scoped component', async () => {
    const { html } = await renderToString('<another-car-detail></another-car-detail>', {
      serializeShadowRoot: true,
      fullDocument: false,
    });
    expect(html).toMatchSnapshot();
  });

  it('supports passing props to components', async () => {
    const { html } = await renderToString(
      '<another-car-detail car=\'{"year":2024, "make": "VW", "model": "Vento"}\'></another-car-detail>',
      {
        serializeShadowRoot: true,
        fullDocument: false,
      },
    );
    expect(html).toMatchSnapshot();
    expect(html).toContain('2024 VW Vento');
  });

  it('supports passing props to components with a simple object', async () => {
    const { html } = await renderToString(`<another-car-detail car=${JSON.stringify(vento)}></another-car-detail>`, {
      serializeShadowRoot: true,
      fullDocument: false,
    });
    expect(html).toMatchSnapshot();
    expect(html).toContain('2024 VW Vento');
  });

  it('does not fail if provided object is not a valid JSON', async () => {
    const { html } = await renderToString(
      `<another-car-detail car='{"year":2024, "make": "VW", "model": "Vento"'></another-car-detail>`,
      {
        serializeShadowRoot: true,
        fullDocument: false,
      },
    );
    expect(html).toContain('<section class="sc-another-car-detail" c-id="1.0.0.0"><!--t.1.1.1.0--> </section>');
  });

  it('supports styles for DSD', async () => {
    const { html } = await renderToString('<another-car-detail></another-car-detail>', {
      serializeShadowRoot: true,
      fullDocument: false,
    });
    expect(html).toContain('section.sc-another-car-detail{color:green}');
  });

  it('only returns the element if we render to DSD', async () => {
    const { html } = await renderToString('<div>Hello World</div>', {
      serializeShadowRoot: true,
      fullDocument: false,
    });
    expect(html).toBe('<div>Hello World</div>');
  });

  it('can render nested components', async () => {
    const { html } = await renderToString(
      `<another-car-list cars=${JSON.stringify([vento, beetle])}></another-car-list>`,
      {
        serializeShadowRoot: true,
        fullDocument: false,
      },
    );
    expect(html).toMatchSnapshot();
    expect(html).toContain('2024 VW Vento');
    expect(html).toContain('2023 VW Beetle');
  });

  it('can render a scoped component within a shadow component', async () => {
    const { html } = await renderToString(`<car-list cars=${JSON.stringify([vento, beetle])}></car-list>`, {
      serializeShadowRoot: true,
      fullDocument: false,
    });
    expect(html).toMatchSnapshot();
    expect(html).toContain(
      '<car-detail class="sc-car-list" custom-hydrate-flag="" c-id="1.2.2.0" s-id="2"><!--r.2--><section class="sc-car-list" c-id="2.0.0.0"><!--t.2.1.1.0-->2024 VW Vento</section></car-detail>',
    );
    expect(html).toContain(
      '<car-detail class="sc-car-list" custom-hydrate-flag="" c-id="1.4.2.0" s-id="3"><!--r.3--><section class="sc-car-list" c-id="3.0.0.0"><!--t.3.1.1.0-->2023 VW Beetle</section></car-detail>',
    );
  });

  it('can render a scoped component within a shadow component (sync)', async () => {
    const input = `<car-list cars=${JSON.stringify([vento, beetle])}></car-list>`;
    const expectedResults = [
      '<car-detail class="sc-car-list" custom-hydrate-flag="" c-id="1.2.2.0" s-id="2"><!--r.2--><section class="sc-car-list" c-id="2.0.0.0"><!--t.2.1.1.0-->2024 VW Vento</section></car-detail>',
      '<car-detail class="sc-car-list" custom-hydrate-flag="" c-id="1.4.2.0" s-id="3"><!--r.3--><section class="sc-car-list" c-id="3.0.0.0"><!--t.3.1.1.0-->2023 VW Beetle</section></car-detail>',
    ] as const;
    const opts = {
      serializeShadowRoot: true,
      fullDocument: false,
    };

    const resultRenderToString = await readableToString(renderToString(input, opts, true));
    expect(resultRenderToString).toContain(expectedResults[0]);
    expect(resultRenderToString).toContain(expectedResults[1]);

    const resultStreamToString = await readableToString(streamToString(input, opts));
    expect(resultStreamToString).toContain(expectedResults[0]);
    expect(resultStreamToString).toContain(expectedResults[1]);
  });

  it('can take over a server side rendered component and re-render it in the browser', async () => {
    const { html } = await renderToString('<cmp-dsd></cmp-dsd>', {
      serializeShadowRoot: true,
      fullDocument: false,
    });

    expect(html).toContain('Count me: 0!');
    const page = await newE2EPage({ html, url: 'https://stencil.com' });
    const button = await page.find('cmp-dsd >>> button');
    await button.click();
    expect(button).toEqualText('Count me: 1!');
  });

  it('can take over a server side rendered component and re-render it in the browser with applied prop', async () => {
    const { html } = await renderToString('<cmp-dsd initial-counter="42"></cmp-dsd>', {
      serializeShadowRoot: true,
      fullDocument: false,
    });

    expect(html).toContain('Count me: 42!');
    const page = await newE2EPage({ html, url: 'https://stencil.com' });
    const button = await page.find('cmp-dsd >>> button');
    await button.click();
    expect(button).toEqualText('Count me: 43!');
  });

  it('can render server side component when client sender renders differently', async () => {
    const { html } = await renderToString('<cmp-server-vs-client></cmp-server-vs-client>', {
      serializeShadowRoot: true,
      fullDocument: false,
    });

    expect(html).toContain('Server vs Client? Winner: Server');
    const page = await newE2EPage({ html, url: 'https://stencil.com' });
    const button = await page.find('cmp-server-vs-client');
    expect(button.shadowRoot.querySelector('div')).toEqualText('Server vs Client? Winner: Client');
  });
});
