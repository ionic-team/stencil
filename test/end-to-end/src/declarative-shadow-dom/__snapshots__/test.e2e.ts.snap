// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToString can render a scoped component within a shadow component 1`] = `"<car-list cars=\\"[{&quot;make&quot;:&quot;VW&quot;,&quot;model&quot;:&quot;Vento&quot;,&quot;year&quot;:2024},{&quot;make&quot;:&quot;VW&quot;,&quot;model&quot;:&quot;Beetle&quot;,&quot;year&quot;:2023}]\\" class=\\"sc-car-list-h\\" custom-hydrate-flag=\\"\\" s-id=\\"1\\"><template shadowrootmode=\\"open\\"><style sty-id=\\"sc-car-list\\">/*!@:host*/.sc-car-list-h{display:block;margin:10px;padding:10px;border:1px solid blue}/*!@ul*/ul.sc-car-list{display:block;margin:0;padding:0}/*!@li*/li.sc-car-list{list-style:none;margin:0;padding:20px}/*!@.selected*/.selected.sc-car-list{font-weight:bold;background:rgb(255, 255, 210)}</style><ul class=\\"sc-car-list\\" c-id=\\"1.0.0.0\\"><li class=\\"sc-car-list\\" c-id=\\"1.1.1.0\\"><car-detail class=\\"sc-car-list\\" custom-hydrate-flag=\\"\\" c-id=\\"1.2.2.0\\" s-id=\\"2\\"><!--r.2--><section class=\\"sc-car-list\\" c-id=\\"2.0.0.0\\"><!--t.2.1.1.0-->2024 VW Vento</section></car-detail></li><li class=\\"sc-car-list\\" c-id=\\"1.3.1.1\\"><car-detail class=\\"sc-car-list\\" custom-hydrate-flag=\\"\\" c-id=\\"1.4.2.0\\" s-id=\\"3\\"><!--r.3--><section class=\\"sc-car-list\\" c-id=\\"3.0.0.0\\"><!--t.3.1.1.0-->2023 VW Beetle</section></car-detail></li></ul></template><!--r.1--></car-list>"`;

exports[`renderToString can render nested components 1`] = `"<another-car-list cars=\\"[{&quot;make&quot;:&quot;VW&quot;,&quot;model&quot;:&quot;Vento&quot;,&quot;year&quot;:2024},{&quot;make&quot;:&quot;VW&quot;,&quot;model&quot;:&quot;Beetle&quot;,&quot;year&quot;:2023}]\\" class=\\"sc-another-car-list-h\\" custom-hydrate-flag=\\"\\" s-id=\\"1\\"><template shadowrootmode=\\"open\\"><style sty-id=\\"sc-another-car-list\\">/*!@:host*/.sc-another-car-list-h{display:block;margin:10px;padding:10px;border:1px solid blue}/*!@ul*/ul.sc-another-car-list{display:block;margin:0;padding:0}/*!@li*/li.sc-another-car-list{list-style:none;margin:0;padding:20px}/*!@.selected*/.selected.sc-another-car-list{font-weight:bold;background:rgb(255, 255, 210)}</style><ul class=\\"sc-another-car-list\\" c-id=\\"1.0.0.0\\"><li class=\\"sc-another-car-list\\" c-id=\\"1.1.1.0\\"><another-car-detail class=\\"sc-another-car-list sc-another-car-detail-h\\" custom-hydrate-flag=\\"\\" c-id=\\"1.2.2.0\\" s-id=\\"2\\"><template shadowrootmode=\\"open\\"><style sty-id=\\"sc-another-car-detail\\">/*!@section*/section.sc-another-car-detail{color:green}</style><section class=\\"sc-another-car-detail\\" c-id=\\"2.0.0.0\\"><!--t.2.1.1.0-->2024 VW Vento</section></template><!--r.2--></another-car-detail></li><li class=\\"sc-another-car-list\\" c-id=\\"1.3.1.1\\"><another-car-detail class=\\"sc-another-car-list sc-another-car-detail-h\\" custom-hydrate-flag=\\"\\" c-id=\\"1.4.2.0\\" s-id=\\"3\\"><template shadowrootmode=\\"open\\"><style sty-id=\\"sc-another-car-detail\\">/*!@section*/section.sc-another-car-detail{color:green}</style><section class=\\"sc-another-car-detail\\" c-id=\\"3.0.0.0\\"><!--t.3.1.1.0-->2023 VW Beetle</section></template><!--r.3--></another-car-detail></li></ul></template><!--r.1--></another-car-list>"`;

exports[`renderToString renders scoped component 1`] = `"<another-car-detail class=\\"sc-another-car-detail-h\\" custom-hydrate-flag=\\"\\" s-id=\\"1\\"><template shadowrootmode=\\"open\\"><style sty-id=\\"sc-another-car-detail\\">/*!@section*/section.sc-another-car-detail{color:green}</style></template><!--r.1--></another-car-detail>"`;

exports[`renderToString supports passing props to components 1`] = `"<another-car-detail car=\\"{&quot;year&quot;:2024, &quot;make&quot;: &quot;VW&quot;, &quot;model&quot;: &quot;Vento&quot;}\\" class=\\"sc-another-car-detail-h\\" custom-hydrate-flag=\\"\\" s-id=\\"1\\"><template shadowrootmode=\\"open\\"><style sty-id=\\"sc-another-car-detail\\">/*!@section*/section.sc-another-car-detail{color:green}</style><section class=\\"sc-another-car-detail\\" c-id=\\"1.0.0.0\\"><!--t.1.1.1.0-->2024 VW Vento</section></template><!--r.1--></another-car-detail>"`;

exports[`renderToString supports passing props to components with a simple object 1`] = `"<another-car-detail car=\\"{&quot;make&quot;:&quot;VW&quot;,&quot;model&quot;:&quot;Vento&quot;,&quot;year&quot;:2024}\\" class=\\"sc-another-car-detail-h\\" custom-hydrate-flag=\\"\\" s-id=\\"1\\"><template shadowrootmode=\\"open\\"><style sty-id=\\"sc-another-car-detail\\">/*!@section*/section.sc-another-car-detail{color:green}</style><section class=\\"sc-another-car-detail\\" c-id=\\"1.0.0.0\\"><!--t.1.1.1.0-->2024 VW Vento</section></template><!--r.1--></another-car-detail>"`;
